### ISSUE TITLE
`plots diff`: fails when checked out file is symlink to directory outside the project

### ISSUE THREAD

**Issue by sjawhar** [iterative/dvc#7668]
# Bug Report

## Description

When using a cache directory outside the project directory (e.g. [shared development server](https://dvc.org/doc/use-cases/fast-data-caching-hub#example-shared-development-server)) with a link cache type, then `diff` commands (at least `plots`, possibly others) can fail with
```
$ dvc plots diff --target plots/sessions/f0960496-1504-46e1-87b7-a268f7623c61/fold_accuracy/cv.csv -- 059b185890a65dab9f588a1c79491762ce94899c HEAD
WARNING: 'plots/sessions/f0960496-1504-46e1-87b7-a268f7623c61/fold_accuracy/cv.csv' was not found at: '059b185890a65dab9f588a1c79491762ce94899c'.
WARNING: 'plots/sessions/f0960496-1504-46e1-87b7-a268f7623c61/fold_accuracy/cv.csv' was not found at: 'HEAD'.
```

The failing logic is here: https://github.com/iterative/dvc/blob/58d960a47ca835acad11fdb813b54524a14e290e/dvc/repo/index.py#L170-L197

Specifically, `out.is_in_repo` evaluates to `False` because the `os.path.realpath(self.fs_path)` resolves to a directory that is not inside `self.repo.root_dir`:
https://github.com/iterative/dvc/blob/58d960a47ca835acad11fdb813b54524a14e290e/dvc/output.py#L374-L387

And so the tree gets populated with the key `("local", "/", "path", "to", "thing")` rather than `("repo", "def", "path", "to", "thing")`, and so [`dvc_fs.repo.index.tree.iteritems(key)` raises a `KeyError`](https://github.com/iterative/dvc/blob/58d960a47ca835acad11fdb813b54524a14e290e/dvc/fs/dvc.py#L114), and so [`not fs.exists(fs_path)`](https://github.com/iterative/dvc/blob/58d960a47ca835acad11fdb813b54524a14e290e/dvc/repo/collect.py#L46), and so the plot is not found.

### Reproduce
```bash
git init
dvc init
dvc plots templates
dvc config cache.dir ../cache
dvc config cache.type symlink
echo 'dvc_plots' > .gitignore
git add .
git commit -m "initial commit"

dvc stage add --plots plot.csv --name plot_thing echo '"a,b\n0,1\n2,3"' \> plot.csv
dvc repro
git add .
git commit -m "First plot commit"

rm plot.csv
dvc stage add --force --plots plot.csv --name plot_thing echo '"a,b\n0,1\n1000,3000"' \> plot.csv
dvc repro
git add .
git commit -m "Second plot commit"

dvc plots diff --target plot.csv -- HEAD~ HEAD
```

* If you do the exact same thing except without `dvc config cache.type symlink`, it works.
* As a workaround, if you delete the plot file from the workspace, it also works. But when the symlink is there, it fails.

### Expected
To produce a `dvc_plots/index.html` with plot at the two revisions

### Environment information

DVC version: 2.10.2 (pip)
---------------------------------
Platform: Python 3.8.10 on Linux-5.13.0-30-generic-x86_64-with-glibc2.29
Supports:
        webhdfs (fsspec = 2022.3.0),
        http (aiohttp = 3.8.1, aiohttp-retry = 2.4.6),
        https (aiohttp = 3.8.1, aiohttp-retry = 2.4.6)
Cache types: hardlink, symlink
Cache directory: ext4 on /dev/mapper/vgubuntu-root
Caches: local
Remotes: None
Workspace directory: ext4 on /dev/mapper/vgubuntu-root
Repo: dvc, git

**Additional Information (if any):**
Verbose output
```
$ dvc plots diff --verbose --target plot.csv -- HEAD~ HEAD
2022-04-30 20:05:45,707 WARNING: 'plot.csv' was not found at: 'HEAD~'.
2022-04-30 20:05:45,707 DEBUG: : ('repo', 'plot.csv')
------------------------------------------------------------
Traceback (most recent call last):
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/fs/dvc.py", line 114, in info
    outs = list(self.repo.index.tree.iteritems(key))
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/data/tree.py", line 92, in iteritems
    for key, (meta, hash_info) in self._trie.iteritems(**kwargs):
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/pygtrie.py", line 633, in iteritems
    node, _ = self._get_node(prefix)
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/pygtrie.py", line 545, in _get_node
    raise KeyError(key)
KeyError: ('repo', 'plot.csv')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/utils/__init__.py", line 497, in wrapper
    vals = func(*args, **kwargs)
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/repo/plots/__init__.py", line 266, in parse
    return _load_sv(path=path, fs=fs, delimiter=",", header=header)
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/repo/plots/__init__.py", line 290, in _load_sv
    with fs.open(path, "r") as fd:
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/fs/base.py", line 186, in open
    return self.fs.open(path, mode=mode, encoding=encoding, **kwargs)
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/fs/repo.py", line 309, in open
    return dvc_fs.open(dvc_path, mode=mode, encoding=encoding, **kwargs)
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/fs/base.py", line 186, in open
    return self.fs.open(path, mode=mode, encoding=encoding, **kwargs)
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/fs/dvc.py", line 77, in open
    fs, fspath = self._get_fs_path(path, **kwargs)
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/fs/dvc.py", line 54, in _get_fs_path
    info = self.info(path)
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/fs/dvc.py", line 116, in info
    raise FileNotFoundError from exc
FileNotFoundError
------------------------------------------------------------
2022-04-30 20:05:45,710 WARNING: 'plot.csv' was not found at: 'HEAD'.
2022-04-30 20:05:45,711 DEBUG: : ('repo', 'plot.csv')
------------------------------------------------------------
Traceback (most recent call last):
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/fs/dvc.py", line 114, in info
    outs = list(self.repo.index.tree.iteritems(key))
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/data/tree.py", line 92, in iteritems
    for key, (meta, hash_info) in self._trie.iteritems(**kwargs):
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/pygtrie.py", line 633, in iteritems
    node, _ = self._get_node(prefix)
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/pygtrie.py", line 545, in _get_node
    raise KeyError(key)
KeyError: ('repo', 'plot.csv')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/utils/__init__.py", line 497, in wrapper
    vals = func(*args, **kwargs)
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/repo/plots/__init__.py", line 266, in parse
    return _load_sv(path=path, fs=fs, delimiter=",", header=header)
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/repo/plots/__init__.py", line 290, in _load_sv
    with fs.open(path, "r") as fd:
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/fs/base.py", line 186, in open
    return self.fs.open(path, mode=mode, encoding=encoding, **kwargs)
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/fs/repo.py", line 309, in open
    return dvc_fs.open(dvc_path, mode=mode, encoding=encoding, **kwargs)
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/fs/base.py", line 186, in open
    return self.fs.open(path, mode=mode, encoding=encoding, **kwargs)
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/fs/dvc.py", line 77, in open
    fs, fspath = self._get_fs_path(path, **kwargs)
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/fs/dvc.py", line 54, in _get_fs_path
    info = self.info(path)
  File "/home/sjawhar/code/.venv/lib/python3.8/site-packages/dvc/fs/dvc.py", line 116, in info
    raise FileNotFoundError from exc
FileNotFoundError
------------------------------------------------------------
DVC failed to load some plots for following revisions: 'HEAD~, HEAD'.
file:///home/sjawhar/code/tmp/dvc_plots/index.html
2022-04-30 20:05:45,712 DEBUG: Analytics is enabled.
2022-04-30 20:05:45,720 DEBUG: Trying to spawn '['daemon', '-q', 'analytics', '/tmp/tmps62l_9jk']'
2022-04-30 20:05:45,721 DEBUG: Spawned '['daemon', '-q', 'analytics', '/tmp/tmps62l_9jk']'
```

**Comment by dtrifiro** [comment]
Possibly related issue on discord, for an user using an NFS mounted cache/remote:
`WARNING: 'plots/cv.csv' was not found at: 'HEAD`

Original discussion:
https://discord.com/channels/485586884165107732/563406153334128681/969131515579609118
https://discord.com/channels/485586884165107732/563406153334128681/969989621960298496

Many thanks to @sjawhar for the detailed report

**Comment by efiop** [comment]
No longer reproducible. Likely solved by related `dvc diff` changes recently. @sjawhar Please give 2.12.0 a try and let us know if you still experience this error. Thank you for the feedback!

**Comment by sjawhar** [comment]
Yep, looks like it's fixed

**Comment by endremborza** [comment]
My issue, that I posted a fix for in #7119 also mentioned here, is still broken in 2.12.0

I was actually redirected to follow this thread. I still manually extend any dvc install I use with the change in #7119

**Comment by dberenbaum** [comment]
@endremborza Thanks for reporting. I reopened #7118 to track that issue.
